File: baseunits/modules/Tumangaonline.pas
unit Tumangaonline;

{$mode objfpc}{$H+}

interface

uses
  Classes, SysUtils, math, WebsiteModules, uData, uBaseUnit, uDownloadsManager,
  XQueryEngineHTML, httpsendthread, synautil, RegExpr;

implementation

const
  apiurl = '/api/v1/';
  apiurlmangas = apiurl + 'mangas';
  apiurlimagenes = apiurl + 'imagenes';
  dirurl = apiurlmangas + '?searchBy=nombre&sortDir=asc&sortedBy=nombre&itemsPerPage=';
  perpage = '1000';
  mangaurl = '/biblioteca/mangas/';
  imgurl = 'https://img1.tumangaonline.com';

function GetDirectoryPageNumber(const MangaInfo: TMangaInformation; var Page: Integer; const WorkPtr: Integer;
  const Module: TModuleContainer): Integer;
begin
  Result := NET_PROBLEM;
  Page := 1;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  MangaInfo.FHTTP.Headers.Values['Referer'] := Module.RootURL;
  MangaInfo.FHTTP.Headers.Values['Cache-Mode'] := 'no-cache';
  if MangaInfo.FHTTP.GET(Module.RootURL + dirurl + '1&page=1') then
  begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
      try
        Page := ceil(StrToIntDef(XPathString('json(*).total'), 1) / StrToInt(perpage));
      finally
        Free;
      end;
  end;
end;

function GetNameAndLink(const MangaInfo: TMangaInformation; const ANames, ALinks: TStringList;
  const AURL: String; const Module: TModuleContainer): Integer;
var
  v: IXQValue;
  s: String;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  MangaInfo.FHTTP.Headers.Values['Referer'] := Module.RootURL;
  MangaInfo.FHTTP.Headers.Values['Cache-Mode'] := 'no-cache';
  if MangaInfo.FHTTP.GET(Module.RootURL + dirurl + perpage + '&page=' + IncStr(AURL)) then
  begin
    Result := NO_ERROR;
    with TXQueryEngineHTML.Create(MangaInfo.FHTTP.Document) do
      try
        for v in XPath('json(*).data()/string-join((id,"/",nombreUrl,codepoints-to-string(10),nombre),"")') do
        begin
          s := v.toString;
          ALinks.Add(mangaurl + SeparateLeft(s, #10));
          ANames.Add(Trim(SeparateRight(s, #10)));
        end;
      finally
        Free;
      end;
  end;
end;

function GetInfo(const MangaInfo: TMangaInformation; const AURL: String;
  const Module: TModuleContainer): Integer;
var
  v, w: IXQValue;
  s, mangaid, purl, num: String;
  p, i: Integer;
begin
  Result := NET_PROBLEM;
  if MangaInfo = nil then Exit(UNKNOWN_ERROR);
  with MangaInfo.mangaInfo, MangaInfo.FHTTP do
  begin
    url := FillHost(Module.RootURL, AURL);
    mangaid := RegExprGetMatch('/\w+/(\d+)/\w+', url, 1);
    if mangaid = '' then Exit;
    Headers.Values['Referer'] := url;
    Headers.Values['Cache-Mode'] := 'no-cache';
    if GET(Module.RootURL + apiurlmangas + '/' + mangaid) then
    begin
      Result := NO_ERROR;
      with TXQueryEngineHTML.Create(Document) do
        try
          coverLink := XPathString('json(*).imageUrl');
          if coverLink <> '' then coverLink := MaybeFillHost(imgurl, coverLink);
          if title = '' then title := XPathString('json(*).nombre');
          status := MangaInfoStatusIfPos(XPathString('json(*).estado'), 'Activo', 'Finalizado');
          summary := XPathString('json(*).info.sinopsis');
          genres := XPathStringAll('json(*).generos().genero');
          AddCommaString(genres, XPathStringAll('json(*).categorias().categoria'));
          artists := XPathStringAll('json(*).artistas().artista');
          authors := XPathStringAll('json(*).autores().autor');
          purl := Module.RootURL + apiurlmangas + '/' + mangaid + '/capitulos?tomo=-1&page=';
          Reset;
          Headers.Values['Referer'] := url;
          Headers.Values['Cache-Mode'] := 'no-cache';
          if GET(purl + '1') then
          begin
            ParseHTML(Document);
            p := StrToIntDef(XPathString('json(*).last_page'), 1);
            for i := 1 to p do
            begin
              if i > 1 then begin
                Reset;
                Headers.Values['Referer'] := url;
                Headers.Values['Cache-Mode'] := 'no-cache';
                if GET(purl + IntToStr(i)) then
                  ParseHTML(Document)
                else
                  Break;
              end;
              for v in XPath('json(*).data()') do
              begin
                mangaid := XPathString('tomo/idManga', v);
                num := XPathString('numCapitulo', v);
                s := v.getProperty('nombre').toString;
                if s = 'null' then s := '';
                if s <> '' then s := ' ' + s;
                s := v.getProperty('numCapitulo').toString + s;
                for w in XPath('jn:members(subidas)', v) do begin
                  chapterLinks.Add(apiurlimagenes + '?idManga=' + mangaid + '&idScanlation=' + XPathString('./idScan', w) + '&numeroCapitulo=' + num + '&visto=true');
                  chapterName.Add(s + ' [' + XPathString('./scanlation/nombre', w) + ']');
                end;
              end;
            end;
            InvertStrings([chapterLinks, chapterName]);
          end;
        finally
          Free;
        end;
    end;
  end;
end;

function GetPageNumber(const DownloadThread: TDownloadThread; const AURL: String;
  const Module: TModuleContainer): Boolean;
var
  s: String;
  v: IXQValue;
begin
  Result := False;
  if DownloadThread = nil then Exit;
  with DownloadThread.FHTTP, DownloadThread.Task.Container do
  begin
    PageLinks.Clear;
    PageNumber := 0;
    Headers.Values['Referer'] := Module.RootURL;
    Headers.Values['Cache-Mode'] := 'no-cache';
    if GET(FillHost(Module.RootURL, AURL)) then
    begin
      Result := True;
      with TXQueryEngineHTML.Create(Document) do
        try
          s := imgurl + '/subidas/' +
            XPathString('json(*)/string-join((capitulo/tomo/idManga,capitulo/numCapitulo,idScan),"/")') + '/';
          for v in XPath('json((json(*).imagenes))()') do
            PageLinks.Add(s + v.toString);
        finally
          Free;
        end;
    end;
  end;
end;

procedure RegisterModule;
begin
  with AddModule do begin
    Website := 'Tumangaonline';
    RootURL := 'https://www.tumangaonline.com';
    Category := 'Spanish';
    Settings.MaxTaskLimit := 1;
    Settings.MaxConnectionLimit := 1;
    OnGetDirectoryPageNumber := @GetDirectoryPageNumber;
    OnGetNameAndLink := @GetNameAndLink;
    OnGetInfo := @GetInfo;
    OnGetPageNumber := @GetPageNumber;
  end;
end;

initialization
  RegisterModule;

end.

File: lua/modules/TuMangaOnline.lua
function getinfo()
  mangainfo.url=MaybeFillHost(module.RootURL, url)
  if http.get(mangainfo.url) then
    local x=TXQuery.Create(http.document)
    if mangainfo.title == '' then
      mangainfo.title = Trim(x.XPathString('//h1[contains(@class, "element-title")]/text()'))
    end
    mangainfo.coverlink = x.xpathstring('//img[contains(@class,"book-thumbnail")]/@src')
    mangainfo.genres=x.xpathstringall('//a[contains(@class, "badge")]')
    mangainfo.authors=Trim(x.xpathstring('//span[@class="list-group-item" and contains(., "Autor")]/a'))
    mangainfo.artists=Trim(x.xpathstring('//span[@class="list-group-item" and contains(., "Artist")]/a'))
    mangainfo.status = MangaInfoStatusIfPos(x.xpathstring('//span[contains(@class, "book-status")]'), 'public', 'final')
    mangainfo.summary = x.xpathstringall('//*[@class="element-description"]/text()', '')
    local v = x.xpath('//div[contains(@class, "chapters")]/ul//li')
    for i = 1, v.count do
      local v1 = v.get(i)
      local name = x.xpathstring('h4', v1)
      local w = x.xpath('div//ul/li/div', v1)
      for j = 1, w.count do
        local w1 = w.get(j)
        local scan = '[' .. x.xpathstring('div[1]', w1) .. ']'
        mangainfo.chapterlinks.add(x.xpathstring('div[contains(@class, "text-right")]/a/@href', w1))
        mangainfo.chapternames.add(name .. ' ' .. scan)
      end
    end
    if mangainfo.chapterlinks.count == 0 then
      local w = x.xpath('//ul[contains(@class, "chapter-list")]/li/div')
      for j = 1, w.count do
        local w1 = w.get(j)
        local scan = '[' .. x.xpathstring('div[1]', w1) .. ']'
        mangainfo.chapterlinks.add(x.xpathstring('div[contains(@class, "text-right")]/a/@href', w1))
        mangainfo.chapternames.add(mangainfo.title .. ' ' .. scan)
      end
    end
    InvertStrings(mangainfo.chapterlinks, mangainfo.chapternames)
    return no_error
  else
    return net_problem
  end
end

function getpagenumber()
  http.headers.values['Referer'] = module.rooturl:gsub('http://', 'https://')
  if not http.get(MaybeFillHost(module.rooturl, url):gsub('http://', 'https://')) then return false; end
  local x = TXQuery.Create(http.Document)
  local u = x.xpathstring('//meta[@property="og:url"]/@content')
  if string.match(u, 'cascade') then
    u = string.gsub(u, '/cascade', '/paginated')
    print(u)
    if not http.get(MaybeFillHost(module.rooturl, u):gsub('http://', 'https://')) then return false; end
    x = TXQuery.Create(http.Document)
  end
  task.pagenumber = tonumber(x.xpathstring('(//select[@id="viewer-pages-select"])[1]/option[last()]/text()'))
  for i = 1, task.pagenumber do
    task.pagecontainerlinks.add(u..'/'..i)
  end
  return true
end

function getimageurl()
  local s = MaybeFillHost(module.rooturl, task.pagecontainerlinks[workid]):gsub('http://', 'https://')
  http.headers.values['Referer'] = module.rooturl:gsub('http://', 'https://')
  if http.get(s) then
    task.pagelinks[workid] = TXQuery.Create(http.document).xpathstring('//img[@class="viewer-image"]/@src')
    
    return true
  end
  return false
end

function getnameandlink()
  local s = '/library?order_item=alphabetically&order_dir=asc&filter_by=title&page='..IncStr(url)
  if http.GET(module.RootURL .. s) then
    local x = TXQuery.Create(http.Document)
    local v = x.xpath('//*[@data-identifier]/a')
    local hasTitles = false
    for i = 1, v.count do
      local v1 = v.get(i)
      links.add(v1.getAttribute('href'))
      names.add(x.xpathstring('div/div[@class="thumbnail-title"]', v1))
      hasTitles = true
    end
    if hasTitles then
      updatelist.CurrentDirectoryPageNumber = updatelist.CurrentDirectoryPageNumber + 1
    end
    return no_error
  else
    return net_problem
  end
end

function Init()
  local m = NewModule()
  m.website = 'Tumangaonline'
  m.rooturl = 'http://tmofans.com' -- Don't set to https because TMO set http for Cloudflare!
  m.category = 'Spanish'
  m.maxtasklimit = 1
  m.maxconnectionlimit = 1
  m.ongetinfo='getinfo'
  m.ongetpagenumber='getpagenumber'
  m.ongetnameandlink='getnameandlink'
  m.ongetimageurl='getimageurl'
end
